src/Sarif/Writers/ResultLogJsonWriter.cs

```diff
@@ -318,18 +318,31 @@ public void CompleteRun()
                 WriteLogicalLocations(_run.LogicalLocations);
             }
 
-            SerializeIfNotNull(_run.Graphs, "graphs");
+            // All ShouldSerialize() will not be triggered automatically when manually write node by node.
+            // To make sure the same logic applied we should call the same method here, if it is defined.
+            if (_run.ShouldSerializeGraphs())
+            {
+                Serialize(_run.Graphs, "graphs");
+            }
 
             // Results go here in schema order
-
-            SerializeIfNotNull(_run.AutomationDetails, "automationDetails");
+            if (_run.ShouldSerializeAutomationDetails())
+            {
+                Serialize(_run.AutomationDetails, "automationDetails");
+            }
```

Comment:
can you create a test using resultlogjsonwriter and verify if it will emit or not.

Reply:
also, check if the other run properties are using the shouldSerialize pattern and if yes, we should add the same logic as this.

Reply:
1. for this one, before this PR, the test we already have
"AnalyzeCommandBase_ShouldEmitAutomationDetailsWhenIdOrGuidExists()"
is already going this path and use the resultlogjsonwriter and verify if it is correctly emitted, (can F5 to step though)

the problem was just there is no case for validate the opposite situation when it should not emit,

so I added cases for null to loop in it, and change the test method name to a generic name reflect it tests both cases.

2. agree, I just go though all fields in the run.cs those have shouldSerialize, and made the same change, updated the PR. the new ones I added are:

ShouldSerializeArtifacts
ShouldSerializeInvocations
ShouldSerializeLogicalLocations